Chapter 6: Optimizing Texture Rendering

AristurtleDev opened the pull request:
This adds two new chapters, a new Chapter 04 and the main part of this PR which is Chapter 06.


Initially this was going to be chapter 5, but I needed to introduce the concept of adding a game library, and it didn't flow right introducing it in the middle of a chapter.  So instead, adding the game library was moved to be chapter 4, making this now chapter 06.

AristurtleDev committed:
Sync with 2d-tutorial-bounty

AristurtleDev committed:
wip

AristurtleDev committed:
WIP: Chapter 05 rework

AristurtleDev committed:
WIP: Split Game Library section into separate chapter, reworked text

AristurtleDev committed:
Adjustments in wording

AristurtleDev committed:
Restructure section to build class up in sections

AristurtleDev committed:
Initial draft of Using Sprite section

AristurtleDev committed:
"when" to "with"

AristurtleDev committed:
reword section

AristurtleDev committed:
Remove tip section, not useful for beginner tutorial

AristurtleDev committed:
Reword section

AristurtleDev committed:
Remove tip, doesn't read right.

AristurtleDev committed:
Wordmark was off center

AristurtleDev committed:
Correct spelling of boundary

AristurtleDev committed:
Fix figure numbers from copy/paste

AristurtleDev committed:
Added conclusion

AristurtleDev committed:
Added Test Your Knowledge questions

AristurtleDev committed:
Remove appendix, was added by accident and not part of this PR

lightalkmst commented on articles/toc.yml:
```diff
@@ -118,8 +118,12 @@
+          href: tutorials/building_2d_games/04_game_library/
+        - name: "05: Working with Textures"
+          href: tutorials/building_2d_games/05_working_with_textures/
+        - name: "06: Optimize Texture Rendering"
```
"optimizing" would be more consistent with the noun forms of the other chapters

AristurtleDev replied:
The actual chapter heading is "Optimizing" so this was a mistake on my part, good catch. Thanks.

lightalkmst commented on articles/toc.yml:
```diff
@@ -118,8 +118,12 @@
+        - name: "05: Working with Textures"
+          href: tutorials/building_2d_games/05_working_with_textures/
+        - name: "06: Optimize Texture Rendering"
+          href: tutorials/building_2d_games/06_optimize_texture_rendering/
```
reminder this line too

AristurtleDev replied:
Got it, thanks

lightalkmst commented on articles/tutorials/building_2d_games/04_game_library/index.md:
```diff
@@ -0,0 +1,126 @@
+2. Enter ".NET New Project" and select that command.
+3. Enter "MonoGame Game Library" and select it as the template to use.
+4. Name the project "MonoGameLibrary".
+5. When prompted for a location, use the default option which will put the new project in a folder next to your game project.
```
i think a comma after option

AristurtleDev replied:
Done

lightalkmst commented on articles/tutorials/building_2d_games/04_game_library/index.md:
```diff
@@ -0,0 +1,126 @@
+2. Choose Add > New Project from the context menu.
+3. Enter "MonoGame Game Library" in the search box, select that template, then click Next.
+4. Name the project "MonoGameLibrary".
+5. The location by default will put the new project in a folder next to your game project, you do not need to adjust this.
```
semicolon instead of comma

AristurtleDev replied:
Done

lightalkmst commented on articles/tutorials/building_2d_games/04_game_library/index.md:
```diff
@@ -0,0 +1,126 @@
+To add the class library using the MonoGame Game Library project template with the dotnet CLI, perform the following:
+
+1. Open a new Command Prompt or Terminal window in the same directory as the *MonoGameSnake.sln* solution file.
+2. Enter the command `dotnet new mglib -n MonoGameLibrary`.  This will create the project and place it in a folder next to your game project.
```
super nitpicky, but all of the other bulleted points are single sentences. semicolon instead of period and second sentence here to be consistent?

AristurtleDev replied:
Wasn't sure about breaking it with semi colons.   I reworked the sentences to be single sentence structures instead

```
1. Open a new Command Prompt or Terminal window in the same directory as the *MonoGameSnake.sln* solution file.
2. Enter the command `dotnet new mglib -n MonoGameLibrary` to create the project, placing it in a folder next to your game project.
3. Enter the command `dotnet sln add ./MonoGameLibrary/` to add the newly created class library project to the *MonoGameSnake.sln* solution file.
```

lightalkmst replied:
that's clean too. lgtm

lightalkmst commented on articles/tutorials/building_2d_games/04_game_library/index.md:
```diff
@@ -0,0 +1,126 @@
+
+1. Open a new Command Prompt or Terminal window in the same directory as the *MonoGameSnake.sln* solution file.
+2. Enter the command `dotnet new mglib -n MonoGameLibrary`.  This will create the project and place it in a folder next to your game project.
+3. Enter the command `dotnet sln add ./MonoGameLibrary/`.  This will add the newly created class library project to the *MonoGameSnake.sln* solution file.
```
this line too

AristurtleDev replied:
See https://github.com/AristurtleDev/docs.monogame.github.io/pull/6#discussion_r1876997735

lightalkmst commented on articles/tutorials/building_2d_games/04_game_library/index.md:
```diff
@@ -0,0 +1,126 @@
+3. Enter the command `dotnet sln add ./MonoGameLibrary/`.  This will add the newly created class library project to the *MonoGameSnake.sln* solution file.
+
+## Adding a Reference To The Class Library
+Now that the game library project has been created, a reference to it needs to be added in our game project.  Without adding a reference, then our game project will be unaware of anything we add to the class library.  To do this:
```
i think remove "then"

AristurtleDev replied:
Yea, it flows better without then. Thanks

lightalkmst commented on articles/tutorials/building_2d_games/04_game_library/index.md:
```diff
@@ -0,0 +1,126 @@
+> The Solution Explorer panel in VSCode is provided by the C# Dev Kit extension that was installed in [Chapter 02](../02_getting_started/index.md#install-the-c-dev-kit-extension).  If you do not see this panel, you can open it by
+>
+> 1. Opening the *Command Palette* (View > Command Palette).
+> 2. Enter "Explorer: Focus on Solution ExplorerView" and select the command.
```
is "ExplorerView" missing a space?

AristurtleDev replied:
It is indeed missing a space

lightalkmst commented on articles/tutorials/building_2d_games/04_game_library/index.md:
```diff
@@ -0,0 +1,126 @@
+1. In the Solution Explorer panel, right-click the *MonoGameSnake* project.
+2. Select Add > Project Reference from the context menu.
+3. Check the box for the *MonoGameLibrary* project.
+4. Click Ok
```
missing period

AristurtleDev replied:
Got it, thanks

lightalkmst commented on articles/tutorials/building_2d_games/04_game_library/index.md:
```diff
@@ -0,0 +1,126 @@
+### dotnet CLI
+To add the game library project as a reference to the game project with the dotnet CLI:
+
+1. Open a new Command Prompt or Terminal window in the same directory as the *MonoGameSnake.csproj* C# project file
```
missing period

AristurtleDev replied:
Got it, thanks

lightalkmst commented on articles/tutorials/building_2d_games/04_game_library/index.md:
```diff
@@ -0,0 +1,126 @@
+Here is a review of what was accomplished in this chapter:
+
+- Learned what a class library is and the benefits of using one.
+- Created a new MonoGame Game Library project
```
missing period

AristurtleDev replied:
Got it, thanks

lightalkmst commented on articles/tutorials/building_2d_games/04_game_library/index.md:
```diff
@@ -0,0 +1,126 @@
+    
+    > Any of the following are advantages of using a class library:
+    > - Reusability: The classes can be easily reused in future game projects by simply adding a reference to the library.
+    > - Organization: Keeps game-specific code separate from reusable library code.
```
start with "it" for consistency

AristurtleDev replied:
Done, thanks.

lightalkmst commented on articles/tutorials/building_2d_games/index.md:
```diff
@@ -19,12 +19,14 @@ This documentation will introduce game development concepts using the MonoGame f
-| [03: The Game1 File](03_the_game1_file/index.md)               | Explore the contents of the Game1 file generated when a new MonoGame game project is created.                                |              |
-| [04: Working with Textures](04_working_with_textures/index.md) | The process of loading image files as textures using the MonoGame content pipeline and rendering them using the SpriteBatch. |              |
+| Chapter                                                                  | Summary                                                                                                                      | Source Files |
+| ------------------------------------------------------------------------ | ---------------------------------------------------------------------------------------------------------------------------- | ------------ |
```
the summaries are kind of grammatically inconsistent with each other as actions or things. pick one

AristurtleDev replied:
Summaries updated to be verbs instead of a mix of verbs and nouns.  Description header in chapters updated to match.

AristurtleDev committed:
Optimize to Optimizing

AristurtleDev committed:
Rephrase to be single sentences for constancy

AristurtleDev committed:
Add missing period

AristurtleDev committed:
Missing period

AristurtleDev committed:
Add "It" for consistency

AristurtleDev committed:
Update image to match new one from chapter 6

AristurtleDev committed:
Missing space

AristurtleDev committed:
Remove "then", not needed

AristurtleDev committed:
Update summaries to be verbs instead of a mix of verbs and nouns

AristurtleDev committed:
Update description metadata of chapters to match summaries

AristurtleDev committed:
Add xref links

SimonDarksideJ commented on articles/tutorials/building_2d_games/04_game_library/index.md:
```diff
@@ -0,0 +1,126 @@
+3.  **Maintainability**: Changes to the library can benefit all games that use it.
+4.  **Testing**: The library code can be tested independently of any specific game.
+
+A class library is a project type that compiles into a Dynamic Link Library (DLL) instead of an executable.  It contains reusable code that can be referenced by other projects, making it perfect for sharing common functionality across multiple games.  MonoGame offers the *MonoGame Game Library* project template that can be used to create a class library.
```
Include a link to the Microsoft reference for what a DLL is :D (for new programmers)

AristurtleDev replied:
Link for MSDN article on DLLs included

SimonDarksideJ commented on articles/tutorials/building_2d_games/04_game_library/index.md:
```diff
@@ -0,0 +1,126 @@
+## Create the Class Library
+Let's use the MonoGame Game Library project template to add a new class library project that will contain our reusable code.
+
+### Visual Studio Code
```
Tabbed view?  rather than have it clutter everything on the screen?

AristurtleDev replied:
Forgot there was the tabbed view in docfx since it's not standard markdown. Switched these to tabs

SimonDarksideJ commented on articles/tutorials/building_2d_games/04_game_library/index.md:
```diff
@@ -0,0 +1,126 @@
+## Adding a Reference To The Class Library
+Now that the game library project has been created, a reference to it needs to be added in our game project.  Without adding a reference, our game project will be unaware of anything we add to the class library.  To do this:
+
+### Visual Studio Code
```
Tabbed view again :D

AristurtleDev replied:
Switched to tabs.

SimonDarksideJ commented on articles/tutorials/building_2d_games/04_game_library/index.md:
```diff
@@ -0,0 +1,126 @@
+To add the game library project as a reference to the game project with the dotnet CLI:
+
+1. Open a new Command Prompt or Terminal window in the same directory as the *MonoGameSnake.csproj* C# project file.
+2. Enter the command `dotnet add reference ../MonoGameLibrary/`.  This will add the *MonoGameLibrary* reference to the *MonoGameSnake* game project.
```
MS recommend using the full syntax to ensure there is no confusion based on where the user is in the folder stack

```cli
dotnet add path/to/project.csproj reference path/to/reference.csproj
```

AristurtleDev replied:
Full syntax for command updated for both sections

SimonDarksideJ commented on articles/tutorials/building_2d_games/04_game_library/index.md:
```diff
@@ -0,0 +1,126 @@
+> [!TIP]
+> These files are needed in more advanced scenarios such as creating a central code base for game logic that is referenced by other projects of which each target different platforms such as desktop, mobile, and console.  Creating a project structure of this type is out of scope for this tutorial.  
+> 
+> If you would like more information on this, Simon Jackson has written the article [Going cross-platform with MonoGame](https://darkgenesis.zenithmoon.com/going-cross-platform-with-monogame.html) which covers this in more detail.
```
That guy is a huckster, wouldn't trust him :P

AristurtleDev replied:
I don't trust him either, he bombards me with documentation reviews 😫

SimonDarksideJ commented on articles/tutorials/building_2d_games/04_game_library/index.md:
```diff
@@ -0,0 +1,126 @@
+> 
+> If you would like more information on this, Simon Jackson has written the article [Going cross-platform with MonoGame](https://darkgenesis.zenithmoon.com/going-cross-platform-with-monogame.html) which covers this in more detail.
+
+## Conclusion
```
One suggestion to address an issue I often see with these kinds of chapters.  It introduces it, adds it, but then doesn't show you how to use it.
Granted this is likely covered later, but by then the reader will need to refer back to this chapter to refresh.

Best to add a section where you add a simple class and then refer to that class from the main project.  Use a class you intend to create to save rework.

Additonally, add a note if they get a warning because the class in the library cannot be found in the main project, reminding them they have forgotten to add a reference :D (you would be surprised how often this comes up :P)

SimonDarksideJ replied:
Maybe also include a screenshot of the final setup in Solution Explorer, so the reader can double check what they see is correct.

AristurtleDev replied:
Section added that include a FramesPerSecondCounter class that can be referenced to ensure the project reference is setup correctly.

AristurtleDev replied:
I keep reading it with the FPS counter class added and it ruins the flow of the documentation by adding and using it just for validation of project reference being setup.  I've simplified the validation step to just adding the using statement and ensuring there is no error.

SimonDarksideJ commented on articles/tutorials/building_2d_games/06_optimizing_texture_rendering/index.md:
```diff
@@ -0,0 +1,503 @@
+description: Explore optimization techniques when rendering textures using a texture atlas.
+---
+
+In [Chapter 05](../05_working_with_textures/index.md), you learned how to load and render textures using [**SpriteBatch**](xref:Microsoft.Xna.Framework.Graphics.SpriteBatch). While rendering individual textures works well for simple games, it can lead to performance issues as your game grows more complex. In this chapter, we'll explore how to optimize texture rendering by reducing texture swaps a ...
```
I know you'll fix this later, but contractions!!! :D

AristurtleDev replied:
Contractions removed

SimonDarksideJ commented on articles/tutorials/building_2d_games/06_optimizing_texture_rendering/index.md:
```diff
@@ -0,0 +1,503 @@
+By the end of this chapter, you'll understand how to organize your game's textures for optimal performance and have a flexible sprite system for your future game projects.
+
+## Texture Swapping
+Every time the [**SpriteBatch.Draw**](xref:Microsoft.Xna.Framework.Graphics.SpriteBatch.Draw(Microsoft.Xna.Framework.Graphics.Texture2D,Microsoft.Xna.Framework.Vector2,System.Nullable{Microsoft.Xna.Framework.Rectangle},Microsoft.Xna.Framework.Color,System.Single,Microsoft.Xna.Framework.Vector2,Microsoft.Xna.Framework.Vector2,Microsoft.Xna.Framework.Graphics.SpriteEffects,System.Single)) method is ...
```
Markdown warning.  Headers MUST have an empty line after them :D

P.S. Use the MSLEarn markdown extension to spot these

AristurtleDev replied:
Added extension and resolved linting warnings such as this

SimonDarksideJ commented on articles/tutorials/building_2d_games/06_optimizing_texture_rendering/index.md:
````diff
@@ -0,0 +1,503 @@
+```
+
+> [!NOTE]
+> The *Sprite.cs* class file is placed in the *MonoGame/Graphics* directory and the class uses the `MonoGame.Graphics` namespace to keep rendering-related classes organized together.  As we add more functionality to the library, we'll continue to use directories and namespaces to maintain a clean structure.
````
Include a link to the MSDN docs on what a namespace is (on the word namespace :D )

AristurtleDev replied:
Link to MSDN article on namespaces added

SimonDarksideJ commented on articles/tutorials/building_2d_games/06_optimizing_texture_rendering/index.md:
```diff
@@ -0,0 +1,503 @@
+You might think that adding these checks are pointless, because when would you ever pass in a null or disposed texture, or provide a source rectangle that is out of bounds of the texture bounds. Of course you would never do this right?  Well, we're all human and sometimes we make mistakes.  It's always best to check yourself to be sure before you publish your game with bugs that could have been a ...
+
+> [!TIP]  
+> Instead of throwing exceptions in the constructor when performing these checks,  `Debug.Assert` is used here.  This has a similar result as throwing an exception, except that the line of code is only ever executed when you run the code in a Debug build.  It asserts that the statement provided is true.  If the statement is false, then code execution will be paused at that line of code similar to ...
```
Include links to the MS source reference when referring to them.

AristurtleDev replied:
Link to namespaces added

AristurtleDev replied:
Links to MSDN article on Debug.Assert added

SimonDarksideJ commented on articles/tutorials/building_2d_games/06_optimizing_texture_rendering/index.md:
```diff
@@ -0,0 +1,503 @@
+However this is not a scalable solution. In a real game with dozens of different textures and complex draw orders for layered sprites, UI elements, particles, etc., managing draw order by texture becomes impractical and will conflict with desired visual layering.
+
+## What is a Texture Atlas
+A texture atlas (also known as a sprite sheet) is a large image file that contains multiple smaller images packed together. Instead of loading separate textures for each game element, you load the single texture file with all the images combined like a scrapbook where all your photos are arranged on the same page.
```
Is it worth adding a note, that most texture atlases are uniformly scaled, meaning that the sprites are of the same dimensions, like for animations (see the platformer sample).  It is less common to have a compact atless due to the issues discussed below :D

Or a comparison of the two approaches.

P.S There should also be a warning about packing, only putting textures that are going to be used on the same screen together.  E.G. Menu textures, Level 1 textures.   While more performant, it is also critical they consider the SIZE of each texture atlas.  Single textures that are too big is also a performance problem to to graphics memory.

AristurtleDev replied:
For now, I'm going to leave this as is and revisit later if needed.

This is a beginner tutorial, so there are some technical limitations, like large texture atlas files, that feel out of scope to a beginner level on boarding tutorial.

Since the game being created is a small snake clone, the texture atlas that will be used for it will contain both the actual snake/food textures and the menu textures as well since it's a relatively small game.  It feels counterintuitive to discuss breaking them up here, then go into creating a single texture atlas that isn't broken up.

SimonDarksideJ reviewed (changes requested):
Good chapters, a few guidance points to add.

But also check the Markdown style guide.  Headers for example MUST have an empty line after them.  VSCode can help with this with the MS Learn plugin, don't ignore the markdown warnings :D

AristurtleDev committed:
Resolve Markdown Lint Warnings:
- MD009/no-trailing-spaces
- MD022/blanks-around-headings
- MD028/no-blanks-blockquote
- MD030/list-marker-space
- MD032/blanks-around-lists

AristurtleDev committed:
Use markdown tabs

AristurtleDev committed:
Include link for MSDN article on DLLs

AristurtleDev committed:
Remove contractions

AristurtleDev committed:
Include link to MSDN article on namespaces

AristurtleDev committed:
Links to MSDN article on Debug.Assert added

AristurtleDev committed:
Use fully syntax for commands

AristurtleDev committed:
Update to use the solution explorer panel

AristurtleDev committed:
Add section validating reference setup correctly

AristurtleDev committed:
Replaced image

dmillot committed:
Fix typo: removed duplicate 'to' (#89)

Custodian19 committed:
Fix Typo in 2_choosing_your_ide_rider.md (#92)

Co-authored-by: Simon (Darkside) Jackson <darkside@zenithmoon.com>

kimmcintyre committed:
chore: Update contribution guidelines (#93)

Co-authored-by: Simon (Darkside) Jackson <darkside@zenithmoon.com>

AristurtleDev committed:
Changed validation section
Instead of adding a class that isn't going to flow in with the documentation, just added a simple validation step instead.

AristurtleDev committed:
Simplify validation step

AristurtleDev committed:
Merge branch 'main' into 2d-tutorial-bounty--chapter-05
